update the governance model to reflect the recent changes (#46)
goharbor/community: GOVERNANCE.md

## Overview

> **Harbor**, a Cloud Native Computing Foundation (CNCF) project, is committed to building an open, inclusive, productive and self-governing open source community focused on building a high-quality cloud native registry. The community is governed by this document with the goal of defining how community should work together to achieve this goal.

Reviewer: Nit: A lot of these changes appear to just remove linebreaks. Is there a reasoning behind this? It makes the diff of ***actual*** changes hard to see. I tend to prefer line breaks because it makes it easier to read in editors that don't have wrapping enabled by defaults.

Author: @nlowe 

Right, remove the line breaks to make it auto warp in the view editor.

Author: @nlowe 

I can keep the line-breaks for easily see the **actual** changes.

Reviewer 2: Line breaks due to vim `textwidth` – sorry. :smile:

### Total Maintainership

Total maintainership is defined as the union of maintainers and core maintainers.

### Maintainers

Nomination for maintainership requires (a) the candidate must have done significant contribution to one of the projects as a contributor, (b) a nomination by an existing maintainer or core maintainer, and (c) an election by a [supermajority](#supermajority) of total maintainership. Likewise, maintainers can be removed by a [supermajority](#supermajority) of the total maintainership or can resign by notifying the core maintainers.

### Core Maintainers

> Nomination for core maintainership requires （a) the candidate must be a maintainer and have already obtained the maintainership for at least **12** months, (b) a nomination by an existing core maintainer, and (c) election by a [supermajority](#supermajority) of the core maintainer group. Likewise, maintainers can be removed by a [supermajority](#supermajority) core maintainer vote or can resign by notifying the core maintainers.

Reviewer 2: Why 12mo? I'm not a fan of putting artificial time requirements here. If someone starts contributing tomorrow and spends 6mo non-stop refactoring huge chunks of code I would not be opposed to elevating the contributor to maintainer and then core maintainer in that time period.

Seems like too much red tape for a small-ish project like Harbor.

Author: Please pay attention, that pending time is not for a contributor to a maintainer, it's for the maintainer to the **core** maintainer.  The maintainer should serve the community as a maintainer role for a reasonable period and then get the ticket to be promoted to core.

### Taking Effect

> The following actions will be applied for maintainership/core maintainership enabling:

Reviewer 2: Please tweak wording; not sure what's meant here.

Reviewer: Perhaps:

```suggestion
Upon adding a new maintainer or core maintainer, the following actions must be performed:
```

Author: @nlowe right, thank u. Will update it later.

* Update the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file
* Add the new maintainer to related mailing groups and social channels if needed
* Add the new maintainer to the [GitHub](https://github.com/orgs/goharbor/teams) organizations and teams if needed
* Publish the announcement to the community

> ### Revoking

Reviewer 2: Again, not a fan of artificial timelines here. Seems superfluous given (a) the size of our project, (b) this has actually never been something we've had to deal with (yet), and (c) "contribution" is not clearly defined here and attempting to do so is likely a deep rabbit hole.

Seems like too much red tape for a small-ish project like Harbor.

Reviewer: Yeah I tend to agree. While I believe the goal appears to be to keep the core maintainers scoped to a group of people who are ***active*** in the project, I don't think we should be doing it with timelines. Who is responsible for tracking maintainer contributions if we do it this way?

Author: So, how do we to handle the case that maintainer is a real maintainer(active) just a shadow one? We really need an active maintainer group.
@clouderati @nlowe

If the maintainer or core maintainer has no contributions in **6** months, the maintainership or core maintainership will be revoked after notifying the maintainer/core maintainer and the community.

### Decision Making

> Ideally, all project decisions are resolved by [consensus](#lazy-consensus). If impossible, any maintainer may call a vote. Unless otherwise specified in this document, any vote will be decided by a [Simple majority](#simple-majority) of the total maintainership, with a requirement of at least one core maintainer voting.

Reviewer 2: consensus != lazy consensus IMHO; would prefer we keep the existing wording in the first sentence.

Is there a specific change that has required a supermajority that has been blocked or slowed down where a simple majority would have been better?

Author: @clouderati NP.

Reviewer: This thread is showing as out-dated due to line ending changes but I think it still needs addressed.

I have yet to see a good argument as to why all votes should be via simple majority unless otherwise stated, especially if the changes granting more power to a large body of developers from a single organization are accepted (as submitted in the present state of the PR).

> Votes by maintainers (either core or non-core) belonging to the same organization cannot exceed the [Simple majority](#simple-majority) threshold. If the number of maintainers (either core or non-core) from the same organization exceeds the simple majority bar, the organization should figure out who will be their representatives to give binding votes in the voting process. The votes from the non-organization-representative maintainers (either core or non-core) will be counted as no-binding votes which will be taken into account when the voting is stalemated (50% VS 50%). The binding and non-binding voting rights will be marked in the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file.

Reviewer 2: Vehement :-1: from me here. I'll post my thoughts on this particular section in the general PR comments.

Author: @clouderati 

Welcome!

> Any changes to the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file will be treated as governance model changes.

Reviewer 2: Tweak wording; e.g.:

> Any changes to the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file are not considered changes to the governance model.

Author: @clouderati 

Totally opposite meanings. The MAINTAINERS.md should be considered as part of the governance model.

### Supermajority

A supermajority is defined as two-thirds of members in the group. A supermajority of [Maintainers](#maintainers), [Core Maintainers](#core-maintainers), or the union of both is required for certain decisions as outlined in the document.

> 

Reviewer: Why was this section removed? I think this is pretty important to prevent one company from having too much sway over the direction of the project.

Author: @nlowe 

Not removed, just changed. Use "votes from the same org cannot exceed the majority (1/2) threshold" to instead the 1 combined organization vote.

Nomination for core maintainership requires (a) a nomination by an existing
core maintainer, and (b) election by a [supermajority](#supermajority) of the core maintainer group. Likewise, maintainers can be removed by a [supermajority](#supermajority) core maintainer vote or can resign by notifying the core maintainers.

### Taking Effect

Upon adding a new maintainer or core maintainer, the following actions must be performed:

* Update the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file
* Add the new maintainer to related mailing groups and social channels if needed
* Add the new maintainer to the [GitHub](https://github.com/orgs/goharbor/teams) organizations and teams if needed
* Publish the announcement to the community

### Decision Making

Ideally, all project decisions are resolved by consensus. If impossible, any
maintainer may call a vote. Unless otherwise specified in this document, any
vote will be decided by a [simple majority](#simple-majority) of the total maintainership, with a requirement of at least one core maintainer voting.

Votes by maintainers (either core or non-core) belonging to the same organization cannot

> exceed the [Simple majority](#simple-majority) threshold. If the number of maintainers

Reviewer 3: `cannot exceed the threshold` means can be equal to the threshold, 50% here for example ? If so, Vmware can make any decisions applying `simple majority`  in my understanding.

Author: @cd1989 

`simple majority` requires > 50%.

Reviewer 2: This is also my understanding (and one of my significant concerns) @cd1989: the updated voting section states that any particular organization / vendor cannot exceed the simple majority, but in this particular case VMware would be able to vote "at" the simple majority resulting in VMware, with the current maintainer landscape, having control over any / all issues that do not require a supermajority.

Reviewer: I think that needs clarified.

@steven-zou For example, if VMware wants to make a change, based on my current understanding of the governance doc, their votes can count for up to 50% of a vote. Assuming all maintainers from VMware vote the same way, is the intention to cause the motion to pass? Or is one more vote from a maintainer other than VMware required?

(either core or non-core) from the same organization exceeds the simple majority bar, the
organization should figure out who will be their representatives to give binding votes in
the voting process. The votes from the non-organization-representative maintainers (either

> core or non-core) will be counted as no-binding votes which will be taken into account when

Reviewer: Should this read as `non-binding`? Can you elaborate what this means and what the intention is here?

the voting is stalemated (50% VS 50%). The binding and non-binding voting rights will be
marked in the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file.

> Any changes to the [MAINTAINERS.md](https://github.com/goharbor/community/blob/master/MAINTAINERS.md) file will be considered as governance model changes.

Reviewer: It is already called out earlier that adding maintainers requires a supermajority vote. Unless something else / additional is trying to be said here I don't think this line needs to be here.

```suggestion
```

* **Accepted**: Proposal is reviewed and accepted (either by consensus or vote).
* **Rejected**: Proposal is reviewed and rejected (either by consensus or vote).

## Other Projects

> The Harbor organization is open to receive new sub-projects under its umbrella. To accept a project into the Harbor organization, it has to meet the following criteria:

Reviewer 2: s/has to/must/

Reviewer: ```suggestion
The Harbor organization is open to receive new sub-projects under its umbrella. To accept a project into the Harbor organization, it must meet the following criteria:
```

> * Must be licensed under the terms of the Apache License v2.0

Reviewer 2: I do not believe we should make this a requirement. Why not GPLv2? MIT? etc.

Reviewer: I agree. According to [https://landscape.cncf.io](https://landscape.cncf.io/license=open-source&project=hosted,graduated,incubating,sandbox), the following licenses are used projects directly related to the CNCF:

![image](https://user-images.githubusercontent.com/794251/52607942-34352f80-2e46-11e9-9930-283c005741de.png)

Why are we limiting this to Apache 2.0?

Author: @clouderati @nlowe 

Refer to the content from https://github.com/fluent/fluentd/blob/master/GOVERNANCE.md and https://github.com/coredns/coredns/blob/master/GOVERNANCE.md.

Reviewer 3: I also think it's unnecessary to limit the license. What's the concern of it @steven-zou ?

Author: @cd1989 @nlowe @steven-zou 

I'm ok with removing the license limitations.

> * Must be related to one or more scopes of the Harbor ecosystem:

Reviewer 2: Tweak wording:

> Must be related to Harbor and its ecosystem, as decided by the core project maintainers.

While I'm generally in favor of being explicit about what we'd accept, this list feels somewhat artificially limited.

Author: OK, np @clouderati

Author: @clouderati 

Refer to the content from https://github.com/fluent/fluentd/blob/master/GOVERNANCE.md and https://github.com/coredns/coredns/blob/master/GOVERNANCE.md.

  * Cloud-native content management
  * Cloud-native content distribution
  * Cloud-native content security
  * Other Cloud-native content registry related

> * Must be supported by a Maintainer not associated or affiliated with the author(s) of the sub-projects

Reviewer 2: Not sure I understand what you're saying here. Would we not want projects created by Harbor maintainers to be accepted as part of the Harbor umbrella? (perhaps I'm misunderstanding)

Reviewer: Yeah I'm confused as well. If this stands, I think we would have had a harder time introducing [hylandsoftware/Harbor.Tagd](https://github.com/hylandsoftware/Harbor.Tagd) and ultimately submitting the proposal that came out of it unless I chose to not join as a maintainer.

Author: @clouderati @nlowe 

Please propose some better words to cover the case that the projects should have enough maintainer/maintainership.

change to: ?   

  Must have at least 2 contributors like Fluentd does?

Refer to the content from Refer to the content from https://github.com/fluent/fluentd/blob/master/GOVERNANCE.md#projects and https://github.com/coredns/coredns/blob/master/GOVERNANCE.md#other-projects .

> * Must have been active for at least **6 months** since it inception

Reviewer 2: Not in favor of artificially imposed timelines. New projects with great potential should be judged as such.

Author: Pend it to hear more comments?
